File: ct_to_shared.py
"""
ct_to_shared.py — Convert Cage-Training trial CSVs to the shared CT/BR schema.

Reads the CSV produced by `CageTrainingDataLoading.m` (`all_trials_<date>.csv`, one
row per trial) and emits a CSV whose columns are the shared variable names.

The MAP now lives in `schema_map.yaml` (the single source of truth). This file only
holds the CT-specific `compute:` transforms (the REGISTRY) and the CLI. The generic
apply logic is in `schema_convert.py`, shared with `br_to_shared.py`.

To change WHICH cage column feeds a shared variable, edit schema_map.yaml.
To change HOW a derived value is computed, edit a function in REGISTRY below.

Not yet validated against real cage data — transforms marked CHECK are unverified.

Usage:
    python ct_to_shared.py --csv all_trials_2026-06-05.csv --task timedelay \
                           --monkey Porthos --out shared_2026-06-05.csv
"""

from __future__ import annotations

import argparse

import numpy as np
import pandas as pd

from schema_convert import apply_side, load_map, _col

ALL_TASKS = ["touch", "touchdot", "touchdotRL", "motion", "timedelay"]


# ----------------------------------------------------------------------------------
# CT compute transforms. Each is (df, ctx) -> Series/array, referenced from the map by
# `ct: { compute: NAME }`. Mark CHECK where the encoding is unconfirmed on real data.
# ----------------------------------------------------------------------------------
def _row_index(df, ctx):
    # 0-based row counter to mirror the BlackRock `index` column.
    return pd.Series(np.arange(len(df)), index=df.index)


def _session(df, ctx):
    # CHECK: a single cage CSV is one date/task folder. Start at session 1 and bump
    # whenever `trialnumber` resets (current < previous). Confirm this matches how you
    # want cage "sessions" defined vs BlackRock's Session join key.
    tn = pd.to_numeric(_col(df, "trialnumber", "session"), errors="coerce")
    return (tn.diff() < 0).cumsum().astype("Int64") + 1


def _direction_sign(df, ctx):
    # `direction` is a bool: True = right (+1), False = left (-1).
    # CHECK: after CSV round-trip the value may be True/False, 1/0, or "true"/"false".
    d = _col(df, "direction", "stimulus_direction")
    return d.map({True: 1, False: -1, 1: 1, 0: -1,
                  "true": 1, "false": -1, "True": 1, "False": -1})


def _stimulus_coherence(df, ctx):
    # CHECK: is cage `coherence` already signed by direction (Notion shows "-76.8 | 76.8"),
    # or a 0-1 magnitude to be signed here? If already signed, change the map to
    # `ct: { direct: coherence }` and delete this function.
    return _col(df, "coherence", "stimulus_coherence") * _direction_sign(df, ctx)


def _choice_rightward(df, ctx):
    # response string -> rightward sign (+1 right, -1 left).
    # CHECK: some cage tasks may use up/down responses; extend this map if so.
    return _col(df, "response", "choice_rightward").map({"right": 1, "left": -1})


def _choice_target(df, ctx):
    # target 1 is DEFINED as the correct target, so accuracy==1 => chose target 1.
    acc = pd.to_numeric(_col(df, "accuracy", "choice_target"), errors="coerce")
    return pd.Series(np.where(acc == 1, 1, np.where(acc == 0, 2, np.nan)), index=df.index)


def _fixation_x(df, ctx):
    # touchdotRL saves an absolute touch position; every other cage task uses screen
    # center (0). BR carries a real Fixation_position_x on its own side.
    if ctx.get("task") == "touchdotRL":
        return _col(df, "xposAbs", "fixation_x")
    return pd.Series(0.0, index=df.index)


def _fixation_y(df, ctx):
    if ctx.get("task") == "touchdotRL":
        return _col(df, "yposAbs", "fixation_y")
    return pd.Series(0.0, index=df.index)


def _is_single_choice(df, ctx):
    # True when only the correct target is shown at choice (so the answer can't be
    # wrong). applies_to = multi_target, so this only runs for motion/timedelay; if you
    # later want single-target tasks to read True, drop applies_to in the map.
    return _col(df, "onlyShowCorrect", "is_single_choice").astype("boolean")


REGISTRY = {
    "row_index": _row_index,
    "session": _session,
    "direction_sign": _direction_sign,
    "stimulus_coherence": _stimulus_coherence,
    "choice_rightward": _choice_rightward,
    "choice_target": _choice_target,
    "fixation_x": _fixation_x,
    "fixation_y": _fixation_y,
    "is_single_choice": _is_single_choice,
}


def convert_cage_to_shared(df: pd.DataFrame, task: str, monkey: str, map_path=None) -> pd.DataFrame:
    entries = load_map(map_path) if map_path else load_map()
    ctx = {"task": task, "monkey": monkey}
    return apply_side(df, entries, side="ct", ctx=ctx, registry=REGISTRY)


def main():
    ap = argparse.ArgumentParser(description="Convert a cage-training trial CSV to the shared schema.")
    ap.add_argument("--csv", required=True, help="input all_trials_<date>.csv from CageTrainingDataLoading")
    ap.add_argument("--task", required=True, choices=ALL_TASKS, help="cage task (from the folder name)")
    ap.add_argument("--monkey", required=True, help="monkey name (from the folder name)")
    ap.add_argument("--out", required=True, help="output CSV path")
    ap.add_argument("--map", default=None, help="path to schema_map.yaml (default: alongside this script)")
    args = ap.parse_args()

    df = pd.read_csv(args.csv)
    shared = convert_cage_to_shared(df, task=args.task, monkey=args.monkey, map_path=args.map)
    shared.to_csv(args.out, index=False)
    print(f"Wrote {len(shared)} rows x {shared.shape[1]} cols -> {args.out}")


if __name__ == "__main__":
    main()

File: ct_to_shared.m
function ct_to_shared(inputCsv, task, monkey, outputCsv, mapPath)
% CT_TO_SHARED  Convert a Cage-Training trial CSV to the shared CT/BR schema.
%
%   ct_to_shared(inputCsv, task, monkey, outputCsv)
%   ct_to_shared(inputCsv, task, monkey, outputCsv, mapPath)
%
% Reads the CSV produced by CageTrainingDataLoading.m (all_trials_<date>.csv,
% one row per trial) and writes a CSV whose columns are the shared variable
% names, filling each from the cage side of the map.
%
% The MAP is schema_map.json (the single source of truth, shared with the Python
% converters in JLab_Python). This file interprets each entry's "ct" spec and
% holds the cage-specific compute transforms (see computeValue below). To change
% WHICH cage column feeds a shared variable, edit schema_map.json. To change HOW a
% derived value is computed, edit computeValue.
%
% `task` is one of: touch | touchdot | touchdotRL | motion | timedelay
% (the cage CSV does not carry the task, so it must be passed in).
%
% Example:
%   ct_to_shared('all_trials_2026-08-12.csv', 'timedelay', 'Porthos', 'shared.csv')
%
% NOTE: not yet validated in MATLAB against real cage data. Transforms with a
% "CHECK" comment are unverified encodings.

    if nargin < 5 || isempty(mapPath)
        mapPath = fullfile(fileparts(mfilename('fullpath')), 'schema_map.json');
    end

    % --- load the cage trials and the map ---
    T = readtable(inputCsv);
    entries = jsondecode(fileread(mapPath));
    if isstruct(entries)              % homogeneous -> struct array; normalise to cell
        entries = num2cell(entries);
    end

    nRows = height(T);
    ctx = struct('task', task, 'monkey', monkey);
    out = table();

    for i = 1:numel(entries)
        e = entries{i};

        % skip optional (derived/measured) rows
        if isfield(e, 'skip') && ~isempty(e.skip) && e.skip
            continue;
        end
        shared = e.shared;

        % applies_to gating: if listed and this task isn't in it -> NaN column
        applies = true;
        if isfield(e, 'applies_to') && ~isempty(e.applies_to)
            at = e.applies_to;
            if ischar(at); at = {at}; end
            applies = any(strcmp(task, at));
        end

        if ~applies
            col = nan(nRows, 1);
        elseif isfield(e, 'ct')
            col = evalSpec(e.ct, T, ctx, nRows, shared);
        else
            col = nan(nRows, 1);
        end

        out.(shared) = col;
    end

    writetable(out, outputCsv);
    fprintf('Wrote %d rows x %d cols -> %s\n', height(out), width(out), outputCsv);
end


% ----------------------------------------------------------------------------------
% Interpret one "ct" spec (a struct with exactly one field naming the rule).
% ----------------------------------------------------------------------------------
function col = evalSpec(spec, T, ctx, nRows, shared)
    if isempty(spec)
        col = nan(nRows, 1); return;
    end
    if isfield(spec, 'direct')
        col = getCol(T, spec.direct, nRows, shared); return;
    end
    if isfield(spec, 'by_task')
        if isfield(spec.by_task, ctx.task)
            col = getCol(T, spec.by_task.(ctx.task), nRows, shared);
        else
            col = nan(nRows, 1);          % this task has no column for the field
        end
        return;
    end
    if isfield(spec, 'const')
        v = spec.const;
        if ischar(v) || isstring(v)
            col = repmat(string(v), nRows, 1);
        else
            col = repmat(v, nRows, 1);
        end
        return;
    end
    if isfield(spec, 'const_ctx')
        col = repmat(string(ctx.(spec.const_ctx)), nRows, 1); return;
    end
    if isfield(spec, 'na')
        col = nan(nRows, 1); return;
    end
    if isfield(spec, 'todo')
        warning('ct_to_shared:todo', '[%s] TODO: %s', shared, spec.todo);
        col = nan(nRows, 1); return;
    end
    if isfield(spec, 'compute')
        col = computeValue(spec.compute, T, ctx, nRows, shared); return;
    end
    warning('ct_to_shared:spec', '[%s] unrecognised spec; filling NaN', shared);
    col = nan(nRows, 1);
end


% ----------------------------------------------------------------------------------
% Cage compute transforms (the "compute: NAME" specs). Mirror ct_to_shared.py.
% ----------------------------------------------------------------------------------
function col = computeValue(name, T, ctx, nRows, shared)
    switch name
        case 'row_index'                     % 0-based row counter
            col = (0:nRows-1)';

        case 'session'                       % CHECK: bump when trialnumber resets
            tn = getNumericCol(T, 'trialnumber', nRows, shared);
            col = cumsum([false; diff(tn) < 0]) + 1;

        case 'direction_sign'                % +1 right, -1 left
            col = directionSign(T, nRows, shared);

        case 'stimulus_coherence'            % CHECK: signed by direction?
            col = getNumericCol(T, 'coherence', nRows, shared) .* directionSign(T, nRows, shared);

        case 'choice_rightward'              % response right -> +1, left -> -1
            col = responseSign(T, nRows, shared);

        case 'choice_target'                 % accuracy 1 -> target 1, 0 -> target 2
            acc = getNumericCol(T, 'accuracy', nRows, shared);
            col = nan(nRows, 1);
            col(acc == 1) = 1;
            col(acc == 0) = 2;

        case 'is_single_choice'              % only the correct target shown
            col = getCol(T, 'onlyShowCorrect', nRows, shared);

        case 'fixation_x'
            col = touchOrConst(T, ctx, nRows, shared, 'xposAbs', 0);
        case 'fixation_y'
            col = touchOrConst(T, ctx, nRows, shared, 'yposAbs', 0);

        case 'target_1_side'                 % correct side tracks direction
            ds = directionSign(T, nRows, shared);
            col = strings(nRows, 1);
            col(ds == 1)  = "right";
            col(ds == -1) = "left";

        case 'target_1_angle'                % right +90, left -90
            col = directionSign(T, nRows, shared) * 90;
        case 'target_2_angle'                % opposite of target 1
            col = directionSign(T, nRows, shared) * -90;

        case 'target_1_x'                    % touchdotRL: xposAbs; motion/timedelay: 0.8/0.2
            if strcmp(ctx.task, 'touchdotRL')
                col = getNumericCol(T, 'xposAbs', nRows, shared);
            elseif any(strcmp(ctx.task, {'motion', 'timedelay'}))
                ds = directionSign(T, nRows, shared);
                col = nan(nRows, 1); col(ds == 1) = 0.8; col(ds == -1) = 0.2;
            else
                col = nan(nRows, 1);
            end
        case 'target_2_x'                    % opposite side of target 1
            ds = directionSign(T, nRows, shared);
            col = nan(nRows, 1); col(ds == 1) = 0.2; col(ds == -1) = 0.8;
        case 'target_1_y'                    % touchdotRL: yposAbs; motion/timedelay: 0
            col = touchOrConst(T, ctx, nRows, shared, 'yposAbs', 0, {'motion', 'timedelay'});
        case 'target_2_y'
            col = zeros(nRows, 1);

        otherwise
            warning('ct_to_shared:compute', '[%s] compute ''%s'' not implemented; filling NaN', shared, name);
            col = nan(nRows, 1);
    end
end


% ----------------------------------------------------------------------------------
% Small helpers.
% ----------------------------------------------------------------------------------
function col = getCol(T, name, nRows, shared)
% Return cage column `name`, or all-NaN (with a warning) if it's missing.
    if ismember(name, T.Properties.VariableNames)
        col = T.(name);
    else
        warning('ct_to_shared:missing', '[%s] source column ''%s'' not found; filling NaN', shared, name);
        col = nan(nRows, 1);
    end
end

function v = getNumericCol(T, name, nRows, shared)
    c = getCol(T, name, nRows, shared);
    if isnumeric(c)
        v = c;
    else
        v = str2double(string(c));
    end
end

function s = toStringCol(c)
% Coerce any column type to a string column (logical/numeric/cell/categorical/string).
    s = string(c);
end

function ds = directionSign(T, nRows, shared)
% `direction` -> +1 (right/true) or -1 (left/false). Robust to bool / 0-1 / "true"/"false".
% CHECK: confirm the encoding readtable produces for `direction` on real data.
    s = toStringCol(getCol(T, 'direction', nRows, shared));
    ds = nan(nRows, 1);
    ds(strcmpi(s, "true")  | s == "1") = 1;
    ds(strcmpi(s, "false") | s == "0") = -1;
end

function col = responseSign(T, nRows, shared)
% response "right" -> +1, "left" -> -1.
% CHECK: some cage tasks may use up/down responses; extend if so.
    s = toStringCol(getCol(T, 'response', nRows, shared));
    col = nan(nRows, 1);
    col(strcmpi(s, "right")) = 1;
    col(strcmpi(s, "left"))  = -1;
end

function col = touchOrConst(T, ctx, nRows, shared, absCol, constVal, tasks)
% touchdotRL -> the absolute touch column; the given tasks (default: all) -> constVal.
    if strcmp(ctx.task, 'touchdotRL')
        col = getNumericCol(T, absCol, nRows, shared);
        return;
    end
    if nargin < 7 || isempty(tasks) || any(strcmp(ctx.task, tasks))
        col = repmat(constVal, nRows, 1);
    else
        col = nan(nRows, 1);
    end
end
